C# 8 Nullable reference types: Define different reference types:  oblivious, nullable, and non-nullable (#1105)
dotnet/csharpstandard: standard/types.md

### 8.2.1 General

> A reference type is a class type, an interface type, an array type, a delegate type, or the `dynamic` type. For each non-nullable reference type, there is a corresponding nullable reference type noted by appending the `?` to the type name.

Reviewer: And a corresponding null-oblivious type? Later on we seem to be treating them as three different types. Let's discuss how we want to describe this, and then try really hard to be consistent.

### 8.3.12 Nullable value types

A ***nullable value type*** can represent all values of its underlying type plus an additional null value. A nullable value type is written `T?`, where `T` is the underlying type. This syntax is shorthand for `System.Nullable<T>`, and the two forms can be used interchangeably.

> Conversely, a ***non-nullable value type*** is any value type other than `System.Nullable<T>` and its shorthand `T?` (for any `T`), plus any type parameter that is constrained to be a non-nullable value type (that is, any type parameter with a value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints))). The `System.Nullable<T>` type specifies the value type constraint for `T`, which means that the underlying type of a nullable value type can be any non-nullable value type. The underlying type of a nullable value type cannot be a nullable value type or a reference type. For example, `int??` and `string?` are invalid types.

Reviewer 2: `string?` is no longer an invalid type, although `Nullable<string>` still is.

### 8.3.12 Nullable value types

A ***nullable value type*** can represent all values of its underlying type plus an additional null value. A nullable value type is written `T?`, where `T` is the underlying type. This syntax is shorthand for `System.Nullable<T>`, and the two forms can be used interchangeably.

> Conversely, a ***non-nullable value type*** is any value type other than `System.Nullable<T>` and its shorthand `T?` (for any `T`), plus any type parameter that is constrained to be a non-nullable value type (that is, any type parameter with a value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints))). The `System.Nullable<T>` type specifies the value type constraint for `T`, which means that the underlying type of a nullable value type can be any non-nullable value type. The underlying type of a nullable value type cannot be a nullable value type or a reference type. For example, `int??` is an invalid types. Nullable reference types are covered in §Types-And-Nullability.

Reviewer: ```suggestion
Conversely, a ***non-nullable value type*** is any value type other than `System.Nullable<T>` and its shorthand `T?` (for any `T`), plus any type parameter that is constrained to be a non-nullable value type (that is, any type parameter with a value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints))). The `System.Nullable<T>` type specifies the value type constraint for `T`, which means that the underlying type of a nullable value type can be any non-nullable value type. The underlying type of a nullable value type cannot be a nullable value type or a reference type. For example, `int??` is an invalid type. Nullable reference types are covered in §Types-And-Nullability.
```

A reference type of the form `T?` (such as `string?`) is an ***annotated reference type***. The annotation `?` indicates the intent that variables of this type are nullable. A reference type of the form `T` (such as `string`) is an ***unannotated reference type***. The absence of the annotation `?` indicates the intent that variables of this type are non-nullable. When the nullable annotation context (§Nullable-Annotation-Context) is enabled, the compiler shall recognize these intents. When the nullable annotation context is disabled, the compiler shall ignore these intents, thereby treating `T` and `T?` in the same (nullable) manner, and generate a warning.

A ***non-nullable reference type*** is a reference type whose variables should not contain the value `null`, and it is an unannotated reference type whose non-null value intent is recognized. A ***nullable reference type*** is any reference type that is not a non-nullable reference type.

> > *Note*: Here, “should” is used rather than “shall,” as this a declaration of intent. The presence of a `?` suffix is not enough to make it a non-nullable type; for that, the annotation context has to be enabled. If it is not, the `?` is ignored, and the type is nullable. *end note*

Reviewer 2: Surely the **presence** of a `?` suffix won't make it a <strong>non-</strong>nullable type, even if the annotation context is enabled.

Author: Fixed.

> Unlike with nullable value types, where value types `V` and `V?` denote different types, given a reference type `R`, the notations `R` and `R?` denote the exact same type; the difference in their notations indicates, at compile time, only the intent of their usage, and allows for static flow analysis. Unlike a nullable value type, a nullable reference type has no relationship to the type `System.Nullable<T>`.

Reviewer 2: The "exact same type" wording feels troublesome.  If the `?` nullable-reference-type annotation is used in an alias 

```csharp
#nullable enable

using ListOfNullableString = System.Collections.Generic.List<string?>;

class C
{
    // warning CS8619
    System.Collections.Generic.List<string> sl = new ListOfNullableString();
}
```

then the standard needs some wording to specify that the initializer of the field `sl` gets both its _type_ and its _nullability annotations_  from the alias `ListOfNullableString`.  Currently, §14.5.2 (Using alias directives) just says "the identifier introduced by the *using_alias_directive* can be used to reference the given namespace or type" and does not mention nullability.

The alternative would be to specify `R` and `R?` as different types that have two-way identity conversions between them, like there are between `object` and `dynamic`, or between similar tuple types.  But perhaps there's something about flow analysis that makes this not possible.

```csharp
#nullable enable

class TypeArgumentDeductionWithIdentityConversion
{
    static void OneNrt(string? a, string b)
    {
        N(ref a, ref b); // warning CS8601
    }
    
    static void TwoNrt(string? a, string? b)
    {
        N(ref a, ref b);
    }
    
    static void ObjectVsDynamic(object[] a, dynamic[] b)
    {
        N(ref a, ref b);
    }
    
    static void TupleElementNames((int X, int Y) a, (int Z, int W) b)
    {
        N(ref a, ref b);
    }
    
    static void N<T>(ref T a, ref T b)
    {
    }
}
```

Author: We're thinking that the identity conversion path is the correct one. (We haven't speced out the warning section yet, but there are other identity conversions that can produce warnings).

- `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, or `bool`.
- Any *enum_type*.
- Any user-defined *struct_type* that is not a constructed type and contains instance fields of *unmanaged_type*s only.
- In unsafe code ([§23.2](unsafe-code.md#232-unsafe-contexts)), any *pointer_type* ([§23.3](unsafe-code.md#233-pointer-types)).

## §Types-And-Nullability Types and nullability

### §Nullable-Types-General General

C# has long supported the notion of a nullable value type; for example, the type `int?` is the nullable version of the value type `int`. (Value types and their nullable variants are described in [§8.3](types.md#83-value-types) and [§8.3.12](types.md#8312-nullable-value-types).) Almost all of the remaining text in this subclause pertains to the addition of support for nullable reference types, with only occasional mention of nullable value types.

There are three forms of nullability for reference types:

- *null-oblivious*: A *null-oblivious* type can be assigned `null`. It's default null state is *not-null*.
- *nullable*: A *nullable-reference-type* can be assigned `null`. It's default null state is *maybe-null*.

> - *non-nullable*" A *non-nullable reference* should not be assigned a `null` value. It's default null state is *not-null*.

Reviewer 2: "Its", not "It's".

- `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, or `bool`.
- Any *enum_type*.
- Any user-defined *struct_type* that is not a constructed type and contains instance fields of *unmanaged_type*s only.
- In unsafe code ([§23.2](unsafe-code.md#232-unsafe-contexts)), any *pointer_type* ([§23.3](unsafe-code.md#233-pointer-types)).

## §Types-And-Nullability Reference Types and nullability

### §Nullable-Types-General General

> Any expression of a reference type can either refer to an object, or be the value `null`, which indicates "no reference". The terms "nullable reference type" and "non-nullable reference type" refer to annotations that declare whether an expression is intended to be null or not. A compiler can provide diagnostics when a expression is not used according to that intent.

Reviewer: ```suggestion
Any expression of a reference type can either refer to an object, or be the value `null`, which indicates "no reference". The terms "nullable reference type" and "non-nullable reference type" refer to annotations that declare whether an expression is intended to permit null values or not. A compiler can provide diagnostics when a expression is not used according to that intent.
```

(Otherwise it sounds like any expression of type `string?` is expected to be null, rather than just null*able*...)

Reviewer: (I'm not overly fond of "permit null values" either - happy to bikeshed on this.)

Author: I accepted the text, and included it in the related PR.

There are three forms of nullability for reference types:

- *null-oblivious*: A *null-oblivious* type can be assigned `null`. It's default null state is *not-null*.

> - *nullable*: A *nullable-reference-type* can be assigned `null`. It's default null state is *maybe-null*.

Reviewer: ```suggestion
- *null-oblivious*: A *null-oblivious* type can be assigned `null`. Its default null state is *not-null*.
- *nullable*: A *nullable-reference-type* can be assigned `null`. Its default null state is *maybe-null*.
```

> - *non-nullable*" A *non-nullable reference* should not be assigned a `null` value. Its default null state is *not-null*.

Reviewer 3: This is referring to “null state” which is not yer defined (it’s ~100 lines further down) – a cross-reference at least should be added.

I also think the default null state difference between *null-oblivious*, which is *not-null*, and *nullable*, which is *maybe-null*, is jarring enough to require an explanation, even if that is just a forward reference.

> Unlike with nullable value types, where value types `V` and `V?` denote different types, given a reference type `R`, the notations `R` and `R?` denote the same underlying type; the difference in their notations indicates, at compile time, only the intent of their usage, and allows for static flow analysis. An identity conversion exists among a nullable reference type, its corresponding non-nullable reference type, and its corresponding null-oblivious reference type (§10.2.2).

Reviewer: We've said they're different types - I think we should reword "same underlying type", but it will require careful wordsmithing. We should see what we say about `dynamic`.

Reviewer 3: @jskeet – Does the above say they are different types? We have:

> The terms "nullable reference type" and "non-nullable reference type" refer to annotations that declare whether an expression is intended to permit null values or not.

This language suggests there is one type which can have different annotations, not that there are multiple distinct types.

Maybe the term “underlying type” should not be overloaded to deal with (non-)nullable reference types? Given the text is leaning on annotations maybe a better term to use here is “non annotated type” (is it *non annotated*, *nonannotated* or *unannotated*?). This would give:

> […] given a reference type `R`, the notations `R` and `R?` denote the same non annotated type […]

> Unlike with nullable value types, where value types `V` and `V?` denote different types, given a reference type `R`, the notations `R` and `R?` denote the same underlying type; the difference in their notations indicates, at compile time, only the intent of their usage, and allows for static flow analysis. An identity conversion exists among a nullable reference type, its corresponding non-nullable reference type, and its corresponding null-oblivious reference type (§10.2.2).

Reviewer 3: > […] and allows for static flow analysis.

Reads wrong, I don’t think the annotations allow static flow analysis *per se*; they may assist/make possible a *particular* static flow analysis.

> > *Note:* By definition, a reference type is nullable; that is, a variable of that type can either contain a reference to an object or be the value `null`, which indicates “no reference.” The disabled nullable context matches the previous standard behavior for reference types. All reference types in these programs are *null-oblivious* reference types. *end note*

Reviewer 3: Forward reference to *nullable context* should be added.

### §Null-oblivious-reference-types Null oblivious reference types

> A *null-oblivious reference type*** is any reference type declared in a disabled nullable annotation context (§Nullable-Annotation-Context). Variables of the type `T` are allowed to be assigned to `null`. The default null-state of a null-oblivious variable is *not-null*.

Reviewer: Is the important part where the type itself is declared, or where some usage is declared? (For example, the nullability of the return type of `public Foo M()` depends on the nullable annotation context of where `M` is declared, not where `Foo` is declared, right?

Reviewer: (Likewise for the sections below.)

A ***non-nullable reference type*** is a reference type declared in an enabled nullable annotation context (§Nullable-Annotation-Context). A variable of a non-nullable reference type of the form `T` should not contain the value `null`. Warnings are generated when a variables of the type `T` is assigned to `null`. The default null-state of a non-nullable variable is *not-null*.

### §Nullable-reference-types Nullable reference types

A reference type of the form `T?` (such as `string?`) is a ***nullable reference type***. The annotation `?` indicates the intent that variables of this type are nullable. When the nullable annotation context (§Nullable-Annotation-Context) is enabled, the compiler shall recognize these intents. When the nullable annotation context is disabled, the compiler shall ignore these intents. When the nullable annotation context is disabled, a reference type of the form `T?` shall declare a null-oblivious reference type.

> *Note:* When the nullable annotation context is disabled, `T` and `T?` both declare null-oblivious reference types. *end note*

### §Nullable-Contexts Nullable contexts

#### §Nullable-Contexts-General General

> Every line of source code has a ***nullable annotation context*** and a ***nullable warning context***. The former controls whether nullable annotations (§Nullable-Annotation-Context) have effect; the latter controls whether nullable warnings (§Nullable-Warning-Context) are issued by the compiler. The state of each of the annotation and warning contexts of a given line is disabled or enabled.

Reviewer 3: I’m not sure about the wording of (*emphasis* added):

> The former controls whether nullable annotations (§Nullable-Annotation-Context) have *effect*; the latter controls whether nullable warnings (§Nullable-Warning-Context) *are issued* by the compiler.

I would think the word “effect” in the context of a language specification would by default be taken to imply some change to the interpretation of the syntax or semantics. NRB is, I understand, being presented as an *optional* analysis which may produce useful comments (aka warnings) about (a fragment of) the language, but *not* have any impact at all on the *meaning* of that (fragment of) the language. In this sense the annotation and context have *no* direct effect at all – they may have an indirect effect in inducing the human/AI who wrote to the code to revise it.

> Both nullable contexts may be specified within source code via nullable directives (§Nullable-Directives) and/or via some implementation-specific mechanism external to the source code. If both approaches are used, nullable directives supersede the settings made via an external mechanism.

Author: This is causing the section renumber failure. That PR should merge first, then this can be updated as needed.

> The default state for the nullable annotation context and the nullable warning context is implementation defined.

Author: Rex had originally written this to say the default state is "disabled". I prefer implementation defined, so in the future, if the implementation changes, it won't require changing the standard. (Yes, it's already documented for roslyn).

> Throughout this specification, all C# code that does not contain nullable directives, or about which no statement is made regarding the current nullable context state, shall be assumed as having been compiled with nullable annotation and nullable warning contexts enabled.

Reviewer: ```suggestion
Throughout this specification, all C# code that does not contain nullable directives, or about which no statement is made regarding the current nullable context state, shall be assumed to have been compiled with nullable annotation and nullable warning contexts enabled.
```

(I think?)

#### §Nullable-Annotation-Context Nullable annotation context

This context controls whether nullable annotations have any effect.

When the nullable annotation context is disabled

> - A variable of any reference type is nullable, so that variable may be initialized with, or assigned a value of, `null`.

Reviewer: All variables, or those declared within the context?

> - No warning shall be generated when a variable of a reference type that possibly has the null value, is dereferenced.

Reviewer: This is weird, given that we've said that the nullable warning context controls whether nullable warnings are issued.

Reviewer: (There's something later that somewhat explains this, but it's quite a bit later.)

- For any reference type `T`, the annotation `?` in `T?` is ignored, as `T` is already a nullable type. An informational message should be generated to that effect.

>   > *Note*: This message is characterized as “informational” rather than “warning,” so as not to confuse it with the state of the nullable warning context, which is unrelated.  *end note*

Reviewer: Is the double space before "end note" deliberate?

>   > *Note*: This message is characterized as “informational” rather than “warning,” so as not to confuse it with the state of the nullable warning context, which is unrelated.  *end note*

Author: nope:

```suggestion
  > *Note*: This message is characterized as “informational” rather than “warning,” so as not to confuse it with the state of the nullable warning context, which is unrelated. *end note*
```

- The null-forgiving operator `!` is ignored.

> *Example*:
>
> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext1", ignoredWarnings:["CS0219","CS8632"], expectedException:"NullReferenceException"} -->
> ```csharp

> > #nullable disable

Reviewer: I think this example would be better if it *just* disabled the nullable annotation context.

> string? s1 = null;    // Informational message; ? is ignored; s2 already nullable
> string s2 = null;     // OK; null initialization of a nullable reference
> s2 = null;            // OK; null assignment to a nullable reference
> char c1 = s2[1];      // OK; no warning on dereference of a possible null; throws NullReferenceException
> c1 = s2![1];          // OK; ! is ignored
> ```
>
> *end example*

When the nullable annotation context is enabled

- For any reference type `T`, the annotation `?` in `T?` makes `T?` a nullable type, whereas the unannotated `T` is non-nullable.
- A warning shall be generated when a variable of an unannotated reference type is initialized with, or assigned a value of, `null`.
- A warning shall be generated when a variable of a nullable or non-nullable reference type, whose value might be `null`, is dereferenced; otherwise, that variable may be dereferenced safely.
- The null-forgiving operator `!` suppresses warnings about dereferencing a possible null reference.

> > *Note*: A Nullable context (§Nullable-Contexts) determines if reference types are considered nullable and non-nullable, or null-oblivious. Switching from a disabled nullable context to an enabled nullable context has the effect of changing the type of reference variables from *null-oblivious* to *non-nullable*. *end note*

Reviewer: Nullable annotation context? (And is the capitalized "N" deliberate?)

<!-- markdownlint-disable MD028 -->

<!-- markdownlint-enable MD028 -->
> *Example*:
>
> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext2", ignoredWarnings:["CS0219"], expectedWarnings:["CS8600","CS8600","CS8602"], expectedException:"NullReferenceException"} -->
> ```csharp
> #nullable enable
> string? s1 = null;    // OK; ? makes s2 nullable
> string s2 = null;     // Warning; s1 is non-nullable
> s2 = null;            // Warning; s1 is non-nullable
> char c1 = s2[1];      // Warning; dereference of a possible null; throws NullReferenceException
> c1 = s2![1];          // The warning is suppressed
> ```
>
> *end example*

#### §Nullable-Warning-Context Nullable warning context

> This context controls whether nullable warnings are issued by the compiler.

Reviewer: Maybe "The nullable warning context controls [...]"? (Same for the annotation context intro.)

> The compiler uses static flow analysis to determine the null state of any reference variable. When the nullable warning context is enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null* or *maybe null*, and

Reviewer: Variable, or expression? (I'm thinking about properties here.) I'll need to look at the notes from before to see what we said about this. Maybe variable is correct, and we'll just add something saying that an implementation can be smarter.

Reviewer: (I see now that we talk about the null state of *expressions* later...)

> The compiler uses static flow analysis to determine the null state of any reference variable. When the nullable warning context is enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null* or *maybe null*, and

Reviewer 3: > The compiler uses static flow analysis […]

Is that a requirement of the Standard? I’m a bit fuzzy on where the (fuzzy) line is here.

As I currently understand an implementation must accept the annotations and pragmas, but may produce any warnings based on those that it likes – specific warnings are not being required.

The hope/allowance is that over time implementations will produce better warnings (and no “hallucinations” to use the in-phrase ;-)) from improving their analyses. But this also implies an implementation could choose to produce no warnings and still be compliant, and if it chooses to do so no analyses would be required…

That the analysis must be *static* could be questioned (consider typing where some limited dynamic typing is supported) but I expect it's unlikely to be an issue for the foreseeable future. But what about it being a *flow* analysis – could a new algorithm be developed that doesn’t fit that category?

TL;DR: I wondering whether, given the lack of requirements to meet conformance for this feature, why the language is not about “an implementation *may*” do such and such, like use a static flow analysis… produce a warning… or even determine that something definitely *is-null* and not just *maybe-null*

> - A warning shall be generated if a reference variable that has been determined to be *maybe null*, is dereferenced.

Reviewer: I think we probably need something about `x.Property.M()` regardless of my comment above - if `Property` is declared to be nullable, there should be *something* about the compiler being able to warn, right?

- The state of a reference variable is *maybe null* unless the compiler can determine one of two conditions:
  1. The variable has been definitely assigned a non-`null` value.
  1. The variable or expression has been checked against `null` before dereferencing it.

> Some warnings are generated regardless of the state of the nullable annotation context. (Examples include throwing a value that maybe `null`, and dereferencing a possibly null reference.) Other warnings are generated only if both the nullable warning context and the nullable annotation context are enabled. (Examples include converting a null literal or possible null value to a non-nullable type, possible null reference assignment, and possible null reference return.)

Reviewer: ```suggestion
Some warnings are generated regardless of the state of the nullable annotation context. (Examples include throwing a value that may be `null`, and dereferencing a possibly null reference.) Other warnings are generated only if both the nullable warning context and the nullable annotation context are enabled. (Examples include converting a null literal or possible null value to a non-nullable type, possible null reference assignment, and possible null reference return.)
```
